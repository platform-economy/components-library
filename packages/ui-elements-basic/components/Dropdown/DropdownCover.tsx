import * as React from 'react';
import styled from 'styled-components';
import { DropdownContextType } from './DropdownContext';
import { withDropdownContext } from './withDropdownContext';
import classnames from 'classnames';

export type DropdownCoverProps = {
  className?: string;
  dropdownContext: DropdownContextType;
  disableClick?: boolean;
  children?: React.ReactNode;
};

class DropdownCoverPure extends React.Component<DropdownCoverProps> {
  constructor(props: DropdownCoverProps) {
    super(props);
  }

  get ref() {
    return this.props.dropdownContext
      ? this.props.dropdownContext.coverRef
      : /* istanbul ignore next */ null;
  }

  get icon() {
    const { opened, iconOpened, iconClosed } = this.props.dropdownContext;
    return opened ? iconOpened : iconClosed;
  }

  render() {
    const { className, dropdownContext, disableClick, children } = this.props;
    const { onCoverClick, onMouseOver, onMouseOut, onFocus, onBlur } = dropdownContext;
    const icon = this.icon;
    return (
      <div
        className={classnames('DropdownCover', className)}
        ref={this.ref as React.RefObject<HTMLDivElement>}
        style={{}}
        onClick={disableClick ? undefined : onCoverClick}
        onMouseEnter={onMouseOver}
        onMouseLeave={onMouseOut}
        onFocus={onFocus}
        onBlur={onBlur}
      >
        {icon ?
          (
            <div className="inlineWrapper">
              <div className="cover">
                {children}
              </div>
              <div className="icon">
                {icon}
              </div>
            </div>
          ) : (
            <div className="cover">
              {children}
            </div>
          )}
      </div>
    );
  }
}

export const DropdownCoverStyled = styled(DropdownCoverPure)``;

export const DropdownCover = withDropdownContext('DropdownCover')(
  DropdownCoverStyled as React.ComponentType<DropdownCoverProps>,
);
