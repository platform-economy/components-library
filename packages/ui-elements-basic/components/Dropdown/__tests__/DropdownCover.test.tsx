import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import 'jest-styled-components';
import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faCaretDown,
  faCaretUp,
} from '@fortawesome/free-solid-svg-icons';
import { DropdownContextType } from '../index';
import { DropdownCoverStyled } from '../DropdownCover';

library.add(
  faCaretDown,
  faCaretUp,
);
enzyme.configure({ adapter: new Adapter() });

const iconOpened = <FontAwesomeIcon icon="caret-up"/>;
const iconClosed = <FontAwesomeIcon icon="caret-down"/>;

const createContext = ({ opened = false } = {}) => {
  return {
    opened,
    iconOpened,
    iconClosed,
    coverRef: React.createRef(),
    onCoverClick: jest.fn(),
    onContentClick: jest.fn(),
    onMouseOver: jest.fn(),
    onMouseOut: jest.fn(),
    onFocus: jest.fn(),
    onBlur: jest.fn(),
  } as DropdownContextType;
};

describe('ui-elements-basic', () => {
  describe('Components', () => {
    describe('DropdownCover', () => {
      [false, true].forEach((opened) => {
        it(`should render (opened: ${opened})`, () => {
          const context = createContext({ opened });
          const tree = TestRenderer
            .create((
              <DropdownCoverStyled dropdownContext={context}>
                <div/>
              </DropdownCoverStyled>
            ));
          expect(tree.toJSON()).toMatchSnapshot();
        });
      });

      it('should attach coverRef to context', () => {
        const context = createContext();
        TestRenderer
          .create((
            <DropdownCoverStyled dropdownContext={context}>
              <div/>
            </DropdownCoverStyled>
          ));
        expect(context.coverRef.current).toBeDefined();
      });

      it('should handle click', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.find('div.test-cover').simulate('click');
        expect(context.onCoverClick).toBeCalled();
      });

      it('should not handle click if disabled', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context} disableClick={true}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.find('div.test-cover').simulate('click');
        expect(context.onCoverClick).toBeCalledTimes(0);
      });

      it('should handle mouse enter', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.simulate('mouseenter');
        expect(context.onMouseOver).toBeCalled();
      });

      it('should handle mouse leave', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.simulate('mouseleave');
        expect(context.onMouseOut).toBeCalled();
      });

      it('should handle focus', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.simulate('focus');
        expect(context.onFocus).toBeCalled();
      });

      it('should handle blur', () => {
        const context = createContext();
        const tree = enzyme
          .mount((
            <DropdownCoverStyled dropdownContext={context}>
              <div className="test-cover"/>
            </DropdownCoverStyled>
          ));
        tree.simulate('blur');
        expect(context.onBlur).toBeCalled();
      });
    });
  });
});
