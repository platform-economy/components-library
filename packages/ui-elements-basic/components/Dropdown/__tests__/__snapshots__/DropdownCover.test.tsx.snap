// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-elements-basic Components DropdownCover should render (opened: false) 1`] = `
<div
  className="DropdownCover "
  onBlur={[MockFunction]}
  onClick={[MockFunction]}
  onFocus={[MockFunction]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
  style={Object {}}
>
  <div
    className="inlineWrapper"
  >
    <div
      className="cover"
    >
      <div />
    </div>
    <div
      className="icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-caret-down fa-w-10 "
        data-icon="caret-down"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`ui-elements-basic Components DropdownCover should render (opened: true) 1`] = `
<div
  className="DropdownCover "
  onBlur={[MockFunction]}
  onClick={[MockFunction]}
  onFocus={[MockFunction]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
  style={Object {}}
>
  <div
    className="inlineWrapper"
  >
    <div
      className="cover"
    >
      <div />
    </div>
    <div
      className="icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-caret-up fa-w-10 "
        data-icon="caret-up"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
